load handling: only launch redis pods if crawler pods successfully loaded.

ikreymer opened the pull request:
~~When crawler pods are stuck in pending (eg. waiting for additional resources), scale down Redis StatefulSet to 0 to avoid using up resources for an unused redis instance.~~

Modified slightly to not launch redis pods until needed:
- waits for crawler pod to be available before starting redis pods, to avoid situation where many crawler pods are in pending mode, but redis pods are still running.
- redis statefulset starts at scale of 0
- once crawler pod becomes available, redis sts is scaled to 1 (via initRedis=true status)
- can happen during 'starting' state or 'waiting_capacity' state
- set to 'running' state only after redis and at least one crawler pod is available
- if no crawler pods available after running, or, if stuck in starting for >60 seconds, switch to 'waiting_capacity' state
- when entering 'waiting_capacity', also scale down redis to 0, wait for crawler pod to become available, then scale up redis to 1, and get back to 'running'

 Fixes #1006

ikreymer committed:
operator: load handling:
- if all crawls pods are pending, then scale down redis sts to 0 to avoid using up memory while no crawls are running
- if crawl pods become available again, ensure redis sts is scaled back to 1

ikreymer committed:
also consider 'ContainerCreating' state as running, as container will not be blocked by resource limits

tw4l reviewed (approved):
Code looks good!

ikreymer commented on backend/btrixcloud/operator.py:
```diff
@@ -232,6 +232,7 @@ async def sync_crawls(self, data: MCSyncData):
         params["force_restart"] = spec.get("forceRestart")
 
         params["redis_url"] = redis_url
+        params["redis_scale"] = 1 if status.state != "waiting_capacity" else 0
```
```suggestion
        params["redis_scale"] = 1 if status.state not in ("starting", "waiting_capacity") else 0
```

If we set it to this, then redis doesn't actually start until crawler pod is creating/running - when crawls are not queued, this does add ~10 seconds to crawl run time, as it first checks that crawler pod is running. When at capacity, does prevent initial burst of redis pods starting up..

ikreymer committed:
additional tweaks:
- don't start redis until crawler is running
- resync after 3 seconds when waiting for redis or in waiting for infra

tw4l commented on backend/btrixcloud/operator.py:
```diff
@@ -461,6 +469,8 @@ async def sync_crawl_state(self, redis_url, crawl, status, pods):
             )
 
+            # resync after 3 seconds
+            status.resyncAfter = 3
```
Looks good - only question is whether we should use a constant here, or make it configurable?

ikreymer replied:
Refactored to make configurable!

ikreymer committed:
op: make fast retry seconds configurable via 'operator_fast_resync_secs' config value,
default to 3 seconds

ikreymer committed:
set_state: if not updating state, ensure state reflects actual value in db

ikreymer committed:
rename new_state -> actual_state

ikreymer committed:
opepartor: crawljob state:
- add 'initRedis' flag to crawl state to indicate where the redis pod should be started
- flag set to true from 'starting' or 'waiting_capacity' to indicate that crawl should not start
- only mark crawl as 'running' if redis is available after 'initRedis' has been set to true

ikreymer commented:
Added additional login, including an `initRedis` flag to the crawljob state. The issue is that want to only mark crawl as a `running` after it is really running. Adding an `initRedis` flag to indicate that the redis pod should now be launched, from either `starting` or `waiting_capacity`, which happens if the crawler pod has successfully started. Set to `running` only when redis and crawler pods are both available, similar to before. The idea is that when set to `running`, the user is guaranteed that crawl is running and not waiting for anything.

ikreymer committed:
typo fix

ikreymer committed:
fix typos
